scanpy/tools/_rank_genes_groups.py

```diff
@@ -18,7 +18,308 @@
+    partition = np.argpartition(scores, -n_top)[-n_top:]
+    partial_indices = np.argsort(scores[partition])[::-1]
+    global_indices = reference_indices[partition][partial_indices]
+
+    return global_indices
+
+
+def _t_test(
+    X,
+    var_names,
+    method,
+    corr_method,
+    expm1_func,
+    rankby_abs,
+    n_genes_user,
+    groups_order,
+    groups_masks,
+    ireference,
+):
+    from scipy import stats
+    from statsmodels.stats.multitest import multipletests
+
+    n_genes = X.shape[1]
+    n_groups = groups_masks.shape[0]
+    d = dict(scores={}, names={}, logfoldchanges={}, pvals={}, pvals_adj={})
+
+    # loop over all masks and compute means, variances and sample numbers
+    means = np.zeros((n_groups, n_genes))
+    vars = np.zeros((n_groups, n_genes))
+    ns = np.count_nonzero(groups_masks, axis=1)
+
+    for imask, mask in enumerate(groups_masks):
+        means[imask], vars[imask] = _get_mean_var(X[mask])
+
+    # test each either against the union of all other groups or against a
+    # specific group
+    for igroup, group_name in enumerate(groups_order):
+        if ireference is None:
+            mask_rest = ~groups_masks[igroup]
+        else:
+            if igroup == ireference:
+                continue
+            else:
+                mask_rest = groups_masks[ireference]
+        mean_group, var_group = means[igroup], vars[igroup]
+        mean_rest, var_rest = _get_mean_var(X[mask_rest])
+
+        # number of observations in group
+        ns_group = ns[igroup]
+        if method == 't-test':
+            ns_rest = np.count_nonzero(mask_rest)
+        elif method == 't-test_overestim_var':
+            # hack for overestimating the variance for small groups
+            ns_rest = ns_group
+        else:
+            raise ValueError('Method does not exist.')
+
+        # TODO: Come up with better solution. Mask unexpressed genes?
+        # See https://github.com/scipy/scipy/issues/10269
+        with np.errstate(invalid="ignore"):
+            scores, pvals = stats.ttest_ind_from_stats(
+                mean1=mean_group,
+                std1=np.sqrt(var_group),
+                nobs1=ns_group,
+                mean2=mean_rest,
+                std2=np.sqrt(var_rest),
+                nobs2=ns_rest,
+                equal_var=False,  # Welch's
+            )
+
+        # Fold change
+        foldchanges = (expm1_func(mean_group) + 1e-9) / (
+            expm1_func(mean_rest) + 1e-9
+        )  # add small value to remove 0's
+
+        # I think it's only nan when means are the same and vars are 0
+        scores[np.isnan(scores)] = 0
+        # This also has to happen for Benjamini Hochberg
+        pvals[np.isnan(pvals)] = 1
+
+        if corr_method == 'benjamini-hochberg':
+            _, pvals_adj, _, _ = multipletests(pvals, alpha=0.05, method='fdr_bh')
+        elif corr_method == 'bonferroni':
+            pvals_adj = np.minimum(pvals * n_genes, 1.0)
```

Comment:
Instead of doing this in the testing function (i.e. `_t_test`, `_wilcoxon`, etc.), could this be done afterwards since it's the same regardless of test type?

This would also let us remove the `corr_method` argument from these functions.